File: c_progs/MatrixMult.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <sys/types.h>
#include <unistd.h>
#include <limits.h>
#include <linux/limits.h>
#include <errno.h>
#include <sys/stat.h>

void create_directory(const char *dir_name) {
    // Attempt to create the directory
    if (mkdir(dir_name, 0755) == -1) {
        // Check if the error is that the directory already exists
        if (errno == EEXIST) {
            printf("Directory '%s' already exists.\n", dir_name);
        } else {
            // Print error message for other errors
            perror("Error creating directory");
        }
    } else {
        printf("Directory '%s' created successfully.\n", dir_name);
    }
}

void write_pid_to_file(const char *filename) {
    const char *directory_name = "c_progs";
    create_directory(directory_name);

    // Open the file for writing; create it if it doesn't exist
    FILE *file = fopen(filename, "w");
    if (file == NULL) {
        perror("Failed to open file");
        return;
    }

    // Get the PID of the current process
    pid_t myPid = getpid();

    // Write the PID to the file, overwriting any previous content
    fprintf(file, "%d\n", myPid);

    // Close the file
    fclose(file);
}


// Function to print a matrix
void printMatrix(long **M, int rowSize, int colSize) {
    for (int i = 0; i < rowSize; i++) {
        for (int j = 0; j < colSize; j++) {
            printf("%ld ", M[i][j]);
        }
        printf("\n");
    }
}

// Function to multiply two matrices A and B
void multiplyMatrix(int row1, int col1, long **A, int row2, int col2, long **B) {
    // Check if multiplication is possible
    if (col1 != row2) {
        printf("Multiplication Not Possible\n");
        return;
    }

    // Allocate memory for result matrix C
    long **C = (long **)malloc(row1 * sizeof(long *));
    for (int i = 0; i < row1; i++) {
        C[i] = (long *)calloc(col2, sizeof(long));
    }

    // Multiply matrices
    for (int i = 0; i < row1; i++) {
        for (int j = 0; j < col2; j++) {
            for (int k = 0; k < col1; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }

    // Print the resultant matrix
    //printf("\nResultant Matrix:\n");
    //printMatrix(C, row1, col2);

    // Free allocated memory for matrix C
    /*for (int i = 0; i < row1; i++) {
        free(C[i]);
    }
    free(C);*/
}

// Function to create a matrix with either random or incremental values
long **createMatrix(int rows, int cols, int randomFactor) {
    long **matrix = (long **)malloc(rows * sizeof(long *));
    for (int i = 0; i < rows; i++) {
        matrix[i] = (long *)malloc(cols * sizeof(long));
    }

    srand(time(NULL));
    int value = 1;

    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            matrix[i][j] = randomFactor ? rand() : value++;
        }
    }
    return matrix;
}

// Main function
int main(int argc, char *argv[]) {
    int val = 1000;
    int row1 = val, col1 = val, row2 = val, col2 = val;

    // Create matrices A and B
    long **A = createMatrix(row1, col1, 1);
    long **B = createMatrix(row2, col2, 1);

    pid_t pid = atoi(argv[1]);
    write_pid_to_file("c_progs/pidfile.txt");
    //sleep(1);
    if (kill(pid, SIGUSR1) == -1) {
        perror("Failed to send SIGUSR1");
        return 1;
    }
    
    // Multiply matrices A and B
    multiplyMatrix(row1, col1, A, row2, col2, B);
    if (kill(pid, SIGUSR2) == -1) {
        perror("Failed to send SIGUSR2");
        return 1;
    }

    // Free allocated memory for matrices A and B
    for (int i = 0; i < row1; i++) {
        free(A[i]);
    }
    free(A);

    for (int i = 0; i < row2; i++) {
        free(B[i]);
    }
    free(B);

    return 0;
}

File: java_progs/MatrixMult.java
package java_progs;

import java.io.IOException;
import java.util.Random;

public class MatrixMult {

    static void printMatrix(long M[][], int rowSize,
                            int colSize)
    {
        for (int i = 0; i < rowSize; i++) {
            for (int j = 0; j < colSize; j++)
                System.out.print(M[i][j] + " ");

            System.out.println();
        }
    }

    // Function to multiply
    // two matrices A[][] and B[][]
    static void multiplyMatrix(int row1, int col1,
                               long A[][], int row2,
                               int col2, long B[][])
    {
        int i, j, k;

        // Print the matrices A and B
        //System.out.println("\nMatrix A:");
        //printMatrix(A, row1, col1);
        //System.out.println("\nMatrix B:");
        //printMatrix(B, row2, col2);

        // Check if multiplication is Possible
        if (row2 != col1) {

            System.out.println(
                "\nMultiplication Not Possible");
            return;
        }

        // Matrix to store the result
        // The product matrix will
        // be of size row1 x col2
        long C[][] = new long[row1][col2];

        // Multiply the two matrices
        for (i = 0; i < row1; i++) {
            for (j = 0; j < col2; j++) {
                for (k = 0; k < row2; k++)
                    C[i][j] += (long)A[i][k] * B[k][j];
            }
        }

        // Print the result
        System.out.println("\nResultant Matrix:");
        //printMatrix(C, row1, col2);
    }

    // Function to create a matrix with either random or incremental values
    static long[][] createMatrix(int rows, int cols, boolean randomFactor) {
        long[][] matrix = new long[rows][cols];
        Random rand = new Random();
        int value = 1;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = randomFactor ? rand.nextLong() : value++;
            }
        }
        return matrix;
    }

    public static void main(String[] args) throws IOException
    {
        int val = 1_000;
        int row1 = val, col1 = val, row2 = val, col2 = val;

        /*int A[][] = { { 1, 1, 1 },
                      { 2, 2, 2 },
                      { 3, 3, 3 },
                      { 4, 4, 4 } };

        int B[][] = { { 1, 1, 1, 1 },
                      { 2, 2, 2, 2 },
                      { 3, 3, 3, 3 } };*/

        long A[][] = createMatrix(row1, col1, true);
        long B[][] = createMatrix(row2, col2, true);
        WritePid.writePidToFile(Long.toString(ProcessHandle.current().pid()));
        Runtime.getRuntime().exec("kill -USR1 " + args[0]);
        multiplyMatrix(row1, col1, A, row2, col2, B);
        Runtime.getRuntime().exec("kill -USR2 " + args[0]);
    }
}
